### PR TITLE
Metrics for sync phases

### PR DESCRIPTION
**Pull request by pinges**
## PR description
This PR is adding metrics for the duration of the chain download and the world state download when doing snap or checkpoint sync.

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastSyncDownloader.java [hunk 1]
```diff
@@ -142,7 +143,7 @@ public void stop() {
     synchronized (this) {
       if (running.compareAndSet(true, false)) {
         LOG.info("Stopping sync");
-        // Cancelling the world state download will also cause the chain download to be cancelled.
+        // Canceling the world state download will also cause the chain download to be cancelled.
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastSyncDownloader.java:146]: I had to look it up. US tends to favor the single "l" version ("canceling" and "canceled") but both are used. However, the double ll version is used in a few spots so if you want to do a find and replace I won't object but it could be a separate PR

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/worldstate/FastWorldDownloadState.java [hunk 2]
```diff
@@ -71,6 +74,10 @@ public synchronized boolean checkCompletion(final BlockHeader header) {
       // so they can give up waiting.
       notifyAll();
       LOG.info("Finished downloading world state from peers");
+
+      // stop the metrics timer for the world download
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/worldstate/FastWorldDownloadState.java:78]: I feel like this comment could be incorporated into either the log message above or the method call below

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java [hunk 3]
```diff
@@ -205,6 +215,9 @@ public TrailingPeerRequirements calculateTrailingPeerRequirements() {
   public CompletableFuture<Void> start() {
     if (running.compareAndSet(false, true)) {
       LOG.info("Starting synchronizer.");
+
+      overallSyncTimingContext = overallSyncTimer.startTimer();
```
Comment (siladu) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java:219]: Not sure the prometheus metrics are designed to time such a long task, could have implications on memory for example.

Was thinking a metric to record the events "sync started" and "sync finished" would be better, then can derive the duration after the fact, e.g. in the dashboard panel. Also, makes the metrics more flexible in case we want to combine the various events, e.g. for different sync phases.

Comment (siladu) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java:219]: Am starting to wonder if metrics is the right way to do this - maybe parsing the logs with Loki is a better way to go? Might be a bit more effort to set up though. Can you see any long term advantage of metrics over logs?

Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java]: In my opinion it is better to use code (Metrics) than logs, because people might change logs when they are not aware that they are used in Grafana.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/worldstate/FastWorldDownloadState.java (hunk 2 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/worldstate/FastWorldDownloadState.java:79]: this has been removed

File: ethereum/eth/src/jmh/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldStateDownloaderBenchmark.java [hunk 4]
```diff
@@ -120,7 +121,8 @@ public void setUpUnchangedState() {
             syncConfig.getWorldStateMaxRequestsWithoutProgress(),
             syncConfig.getWorldStateMinMillisBeforeStalling(),
             Clock.fixed(Instant.ofEpochSecond(1000), ZoneOffset.UTC),
-            metricsSystem);
+            metricsSystem,
+            SyncDurationMetrics.NO_METRICS_SYNC_DURATION_METRICS);
```
Comment (garyschulte) [inline ethereum/eth/src/jmh/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldStateDownloaderBenchmark.java:125]: there seems to be quite a few of these NO_METRICS_SYNC_DURATION_METRICS, it might be useful to have an overloaded constructor that delegates to the main constructor with NO_METRICS...

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java [hunk 5]
```diff
@@ -205,6 +213,9 @@ public TrailingPeerRequirements calculateTrailingPeerRequirements() {
   public CompletableFuture<Void> start() {
     if (running.compareAndSet(false, true)) {
       LOG.info("Starting synchronizer.");
+
+      syncDurationMetrics.startTimer(SyncDurationMetrics.Labels.TOTAL_SYNC_DURATION);
```
Comment (garyschulte) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java:217]: does startTimer work as intended if it has been stopped?  Is it additive, or does it replace the prior metric values?

Thinking about the worldstate heal re-use of the synchronizer

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java [hunk 6]
```diff
@@ -271,7 +279,8 @@ public synchronized void reloadTrieHeal() {
   }
 
   public synchronized void startFlatDatabaseHeal(final BlockHeader header) {
-    LOG.info("Initiating the healing process for the flat database");
+    syncDurationMetrics.startTimer(SyncDurationMetrics.Labels.FLAT_DB_HEAL);
+    syncDurationMetrics.stopTimer(SyncDurationMetrics.Labels.CHAIN_DOWNLOAD_DURATION);
```
Comment (garyschulte) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:283]: this is a surprising place to find this imo.  If a node starts and does not perform a worldstate sync/heal, will the chain download duration count upwards endlessly?  IIRC the chain download pipeline doesn't ever stop until the node stops.

Perhaps a better place to stop this timer would be as an InitialSyncListener in PipelineChainDownloader.

**Review by garyschulte — COMMENTED** [review]
I had to look up the difference between Summary and Histogram timers, SimpleTimer seems to be a good name for this

Some comments, but generally LGTM

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java (hunk 5 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/DefaultSynchronizer.java:217]: startTimer() only works the first time it is called. After the first start the call has no effect.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java (hunk 6 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:283]: You are right, that is surprising. The problem with the PipelineChainDownloader is that it is started and stopped multiple times. I have made clear in the PR description that these metrics are only for snap and cp sync. Fast sync is broken at the moment and we will probably deprecate it once snap works for private networks.
I have moved the stopTimer out of the startFlatDatabaseHeal, as you can still configure besu to not do it.

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.HashMap;
+
+/**
+ * This class manages the synchronization duration metrics for the Hyperledger Besu project. It
+ * provides methods to start and stop timers for various synchronization phases.
+ */
+public class SyncDurationMetrics {
+
+  /** A {@link SyncDurationMetrics} instance that does not record any metrics. */
+  public static final SyncDurationMetrics NO_METRICS_SYNC_DURATION_METRICS =
```
Comment (siladu) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:31]:
```suggestion
  public static final SyncDurationMetrics NO_OP_SYNC_DURATION_METRICS =
```
?

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+  /**
+   * Creates a new {@link SyncDurationMetrics} instance.
+   *
+   * @param metricsSystem The {@link MetricsSystem} to use to record metrics.
+   */
+  public SyncDurationMetrics(final MetricsSystem metricsSystem) {
+    timer =
+        metricsSystem.createSimpleLabelledTimer(
+            BesuMetricCategory.SYNCHRONIZER, "sync_duration", "Time taken to sync", "name");
```
Comment (siladu) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:46]:
```suggestion
            BesuMetricCategory.SYNCHRONIZER, "total_sync_duration", "Total time taken to sync", "name");
```

Would be nice to state in the description (or at least the comment) exactly what this is measuring in terms of the stages, e.g. chain download, heal, backwards sync, etc

Comment (siladu) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:46]: What is the purpose of the "name" string here?

Comment (siladu) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:46]: hmmm...is "sync_duration" actually a label for a group of metrics?

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  public void stopTimer(final Labels label) {
+    OperationTimer.TimingContext context = timers.get(label.name());
+    if (context != null) {
+      context.stopTimer();
+    }
+  }
+
+  /** Enum representing the different synchronization phases. */
+  public enum Labels {
+    /** Total time taken to get into sync. */
```
Comment (siladu) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:72]: Would be nice to state exactly what this is measuring in terms of the other stages, e.g. chain download, heal, backwards sync, etc. See other comment also

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+   * @param label The synchronization phase to stop the timer for.
+   */
+  public void stopTimer(final Labels label) {
+    OperationTimer.TimingContext context = timers.get(label.name());
+    if (context != null) {
+      context.stopTimer();
+    }
+  }
+
+  /** Enum representing the different synchronization phases. */
```
Comment (siladu) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:70]: Would be helpful to indicate how these various durations might overlap as it could be quite confusing to interpret the metrics without this domain knowledge, see other comment about total_sync_duration.

I think this warrants user documentation as well - maybe a wider issue that explains the sync stages and how these metrics relate, an updated version of https://besu.hyperledger.org/public-networks/how-to/monitor/understand-metrics perhaps

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/PipelineChainDownloaderTest.java [hunk 11]
```diff
@@ -69,13 +70,15 @@ public class PipelineChainDownloaderTest {
   public void setUp() {
     syncTarget = new SyncTarget(peer1, commonAncestor);
     syncTarget2 = new SyncTarget(peer2, commonAncestor);
+    final NoOpMetricsSystem noMetricsSystem = new NoOpMetricsSystem();
```
Comment (siladu) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/PipelineChainDownloaderTest.java:73]:
```suggestion
    final NoOpMetricsSystem noOpMetricsSystem = new NoOpMetricsSystem();
```

File: ethereum/eth/src/jmh/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldStateDownloaderBenchmark.java (hunk 4 above)
Comment (siladu) [reply ethereum/eth/src/jmh/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/WorldStateDownloaderBenchmark.java:125]: Agree.

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java (hunk 8 above)
Comment (pinges) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:46]: actually sync_duration is the actual name of the metric, which then has multiple children, which we identify with a label.

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java (hunk 10 above)
Comment (pinges) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:70]: I agree. I think I should update that for the sync duration timers!

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java (hunk 9 above)
Comment (pinges) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:72]: It says what it is measuring :-) 
Total time to get into sync, From the start of the sync to the last thing that is happening (BWS).

Comment (siladu) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:72]: That implies a knowledge of our sync stages which not everyone who reads this will have, but even then, it doesn't include BWS which is surprising, I think that's worth mentioning at least.

Most important this is explained in the user docs I guess, devs can always refer to that, it's just convenient when you're reading the code IMO

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  public void stopTimer(final Labels label) {
+    OperationTimer.TimingContext context = timers.remove(label.name());
+    if (context != null) {
+      context.stopTimer();
+    }
+  }
+
+  /** Enum representing the different synchronization phases. */
+  public enum Labels {
+    /** Total time taken to get into sync. */
```
Comment (siladu) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:72]:
```suggestion
    /** Total time taken to get into sync. It is useful for SNAP and CHECKPOINT sync-modes only. 
    
    Total sync duration includes the separate stages mentioned below, some of which occur in parallel. 
    
    Total sync duration excludes the backwards sync stage due to implementation challenges. The backwards sync should be a very short duration following the other sync stages.
      */
```

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    TOTAL_SYNC_DURATION,
+    /** Time taken to download the chain data (headers, blocks, receipts). */
+    CHAIN_DOWNLOAD_DURATION,
+    /** Time taken to download the initial world state, before the healing step. */
+    SNAP_INITIAL_WORLD_STATE_DOWNLOAD_DURATION,
+    /** Time taken to heal the world state, after the initial download. */
+    SNAP_WORLD_STATE_HEALING_DURATION,
+    /** Time taken to do the fast world state download. */
+    FAST_WORLD_STATE_DOWNLOAD_DURATION,
```
Comment (siladu) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:81]: unused now?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java [hunk 14]
```diff
@@ -102,60 +108,53 @@ public SnapWorldDownloadState(
...
+    metricsSystem.createLongGauge(
+        BesuMetricCategory.SYNCHRONIZER,
+        "snap_world_state_pending_trie_node_requests_current",
+        "Number of trie node pending requests for snap sync world state download",
+        pendingTrieNodeRequests::size);
+    syncDurationMetrics.startTimer(
+        SyncDurationMetrics.Labels.SNAP_INITIAL_WORLD_STATE_DOWNLOAD_DURATION);
+    LOG.info(
+        "startTimer SNAP_INITIAL_WORLD_STATE_DOWNLOAD_DURATION: {}",
+        LocalDateTime.now(ZoneId.systemDefault()));
```
Comment (pinges) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:157]:
```suggestion
```

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java [hunk 15]
```diff
@@ -271,7 +280,7 @@ public synchronized void reloadTrieHeal() {
   }
 
   public synchronized void startFlatDatabaseHeal(final BlockHeader header) {
-    LOG.info("Initiating the healing process for the flat database");
+    syncDurationMetrics.startTimer(SyncDurationMetrics.Labels.FLAT_DB_HEAL);
```
Comment (pinges) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:283]:
```suggestion
    LOG.info("Initiating the healing process for the flat database");
    syncDurationMetrics.startTimer(SyncDurationMetrics.Labels.FLAT_DB_HEAL);
```

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+      */
+    TOTAL_SYNC_DURATION,
+    /** Time taken to download the chain data (headers, blocks, receipts). */
+    CHAIN_DOWNLOAD_DURATION,
+    /** Time taken to download the initial world state, before the healing step. */
+    SNAP_INITIAL_WORLD_STATE_DOWNLOAD_DURATION,
+    /** Time taken to heal the world state, after the initial download. */
+    SNAP_WORLD_STATE_HEALING_DURATION,
+    /** Time taken to do the fast world state download. */
+    FAST_WORLD_STATE_DOWNLOAD_DURATION,
```
Comment (pinges) [inline metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java:86]:
```suggestion
```

File: metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java (hunk 13 above)
Comment (pinges) [reply metrics/core/src/main/java/org/hyperledger/besu/metrics/SyncDurationMetrics.java]: Done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java]: Done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java [hunk 17]
```diff
@@ -271,7 +280,7 @@ public synchronized void reloadTrieHeal() {
   }
 
   public synchronized void startFlatDatabaseHeal(final BlockHeader header) {
-    LOG.info("Initiating the healing process for the flat database");
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:274]: Done

**Review by siladu — APPROVED** [review]
LGTM - thanks for the changes!
